File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/jvmAndroidMain/kotlin/com/runanywhere/sdk/public/extensions/LLM/RunAnywhereToolCalling.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Public API for tool calling (function calling) with LLMs.
 * Allows LLMs to request external actions (API calls, device functions, etc.)
 *
 * ARCHITECTURE:
 * - CppBridgeToolCalling: C++ bridge for parsing <tool_call> tags (SINGLE SOURCE OF TRUTH)
 * - This file: Tool registration, executor storage, orchestration
 * - Orchestration: generate → parse (C++) → execute → loop
 *
 * *** ALL PARSING LOGIC IS IN C++ (rac_tool_calling.h) - NO KOTLIN FALLBACKS ***
 *
 * Mirrors Swift SDK's RunAnywhere+ToolCalling.swift
 */

package com.runanywhere.sdk.public.extensions.LLM

import com.runanywhere.sdk.foundation.SDKLogger
import com.runanywhere.sdk.foundation.bridge.extensions.CppBridgeToolCalling
import com.runanywhere.sdk.public.RunAnywhere
import com.runanywhere.sdk.public.extensions.generateStream
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Thread-safe tool registry for tool registration and lookup.
 */
private object ToolRegistry {
    private val mutex = Mutex()
    private val tools = mutableMapOf<String, RegisteredTool>()

    suspend fun register(definition: ToolDefinition, executor: ToolExecutor) =
        mutex.withLock {
            tools[definition.name] = RegisteredTool(definition, executor)
        }

    suspend fun unregister(toolName: String) =
        mutex.withLock {
            tools.remove(toolName)
        }

    suspend fun getAll(): List<ToolDefinition> =
        mutex.withLock {
            tools.values.map { it.definition }
        }

    suspend fun get(toolName: String): RegisteredTool? =
        mutex.withLock {
            tools[toolName]
        }

    suspend fun clear() =
        mutex.withLock {
            tools.clear()
        }
}

/**
 * Tool calling extension for RunAnywhere.
 */
object RunAnywhereToolCalling {
    private const val TAG = "ToolCalling"
    private val logger = SDKLogger(TAG)

    // ========================================================================
    // TOOL REGISTRATION
    // ========================================================================

    /**
     * Register a tool that the LLM can use.
     *
     * Tools are stored in-memory and available for all subsequent generateWithTools calls.
     * Executors run in Kotlin and have full access to Kotlin/Android APIs.
     *
     * @param definition Tool definition (name, description, parameters)
     * @param executor Suspend function that executes the tool
     */
    suspend fun registerTool(
        definition: ToolDefinition,
        executor: ToolExecutor,
    ) {
        ToolRegistry.register(definition, executor)
        logger.info("Registered tool: ${definition.name}")
    }

    /**
     * Unregister a tool by name.
     *
     * @param toolName The name of the tool to remove
     */
    suspend fun unregisterTool(toolName: String) {
        ToolRegistry.unregister(toolName)
        logger.info("Unregistered tool: $toolName")
    }

    /**
     * Get all registered tool definitions.
     *
     * @return List of registered tool definitions
     */
    suspend fun getRegisteredTools(): List<ToolDefinition> {
        return ToolRegistry.getAll()
    }

    /**
     * Clear all registered tools.
     */
    suspend fun clearTools() {
        ToolRegistry.clear()
        logger.info("Cleared all registered tools")
    }

    // ========================================================================
    // TOOL EXECUTION
    // ========================================================================

    /**
     * Execute a tool call.
     *
     * Looks up the tool in the registry and invokes its executor with the provided arguments.
     * Returns a ToolResult with success/failure status.
     *
     * @param toolCall The tool call to execute
     * @return Result of the tool execution
     */
    suspend fun executeTool(toolCall: ToolCall): ToolResult {
        val tool = ToolRegistry.get(toolCall.toolName)

        if (tool == null) {
            return ToolResult(
                toolName = toolCall.toolName,
                success = false,
                error = "Unknown tool: ${toolCall.toolName}",
                callId = toolCall.callId,
            )
        }

        return try {
            val result = tool.executor(toolCall.arguments)
            ToolResult(
                toolName = toolCall.toolName,
                success = true,
                result = result,
                callId = toolCall.callId,
            )
        } catch (e: Exception) {
            logger.error("Tool execution failed: ${e.message}")
            ToolResult(
                toolName = toolCall.toolName,
                success = false,
                error = e.message ?: "Unknown error",
                callId = toolCall.callId,
            )
        }
    }

    // ========================================================================
    // GENERATE WITH TOOLS
    // ========================================================================

    /**
     * Generates a response with tool calling support.
     *
     * Orchestrates a generate → parse → execute → loop cycle:
     * 1. Builds a system prompt describing available tools (C++)
     * 2. Generates LLM response
     * 3. Parses output for `<tool_call>` tags (C++ - SINGLE SOURCE OF TRUTH)
     * 4. If tool call found and autoExecute is true, executes and continues
     * 5. Repeats until no more tool calls or maxToolCalls reached
     *
     * @param prompt The user's prompt
     * @param options Tool calling options
     * @return Result containing final text, all tool calls made, and their results
     */
    suspend fun generateWithTools(
        prompt: String,
        options: ToolCallingOptions? = null,
    ): ToolCallingResult {
        // Ensure SDK is initialized
        require(RunAnywhere.isInitialized) { "SDK not initialized" }

        val opts = options ?: ToolCallingOptions()
        val registeredTools = ToolRegistry.getAll()
        val tools = opts.tools ?: registeredTools

        // Build system prompt using C++ (SINGLE SOURCE OF TRUTH)
        val systemPrompt = buildToolSystemPrompt(tools, opts)
        var fullPrompt = if (systemPrompt.isEmpty()) prompt else "$systemPrompt\n\nUser: $prompt"

        val allToolCalls = mutableListOf<ToolCall>()
        val allToolResults = mutableListOf<ToolResult>()
        var finalText = ""

        repeat(opts.maxToolCalls) { iteration ->
            logger.debug("Tool calling iteration $iteration")

            // Generate response
            val responseText = generateAndCollect(fullPrompt, opts.temperature, opts.maxTokens)

            // Parse for tool calls using C++ (SINGLE SOURCE OF TRUTH - NO FALLBACK)
            val (cleanText, toolCall) = CppBridgeToolCalling.parseToolCallToObject(responseText)
            finalText = cleanText

            if (toolCall == null) {
                logger.debug("No tool call found, generation complete")
                return ToolCallingResult(
                    text = finalText,
                    toolCalls = allToolCalls,
                    toolResults = allToolResults,
                    isComplete = true,
                )
            }

            allToolCalls.add(toolCall)
            logger.info("Found tool call: ${toolCall.toolName}")

            if (!opts.autoExecute) {
                return ToolCallingResult(
                    text = finalText,
                    toolCalls = allToolCalls,
                    toolResults = emptyList(),
                    isComplete = false,
                )
            }

            // Execute tool
            val result = executeTool(toolCall)
            allToolResults.add(result)
            logger.info("Tool ${toolCall.toolName} executed: ${if (result.success) "success" else "failed"}")

            // Build follow-up prompt using C++ (SINGLE SOURCE OF TRUTH)
            val toolResultJson =
                CppBridgeToolCalling.toolValueToJsonString(
                    result.result ?: mapOf("error" to ToolValue.string(result.error ?: "Unknown error")),
                )

            fullPrompt =
                CppBridgeToolCalling.buildFollowupPrompt(
                    originalPrompt = prompt,
                    toolsPrompt = if (opts.keepToolsAvailable) CppBridgeToolCalling.formatToolsForPrompt(tools, opts.format) else null,
                    toolName = toolCall.toolName,
                    toolResultJson = toolResultJson,
                    keepToolsAvailable = opts.keepToolsAvailable,
                )
        }

        return ToolCallingResult(
            text = finalText,
            toolCalls = allToolCalls,
            toolResults = allToolResults,
            isComplete = true,
        )
    }

    /**
     * Continue generation after manual tool execution.
     *
     * Use this when autoExecute is false. After receiving a ToolCallingResult
     * with isComplete: false, execute the tool yourself, then call this to continue.
     *
     * @param previousPrompt The original user prompt
     * @param toolCall The tool call that was executed
     * @param toolResult The result of executing the tool
     * @param options Tool calling options for the continuation
     * @return Result of the continued generation
     */
    suspend fun continueWithToolResult(
        previousPrompt: String,
        toolCall: ToolCall,
        toolResult: ToolResult,
        options: ToolCallingOptions? = null,
    ): ToolCallingResult {
        val resultJson =
            CppBridgeToolCalling.toolValueToJsonString(
                toolResult.result ?: mapOf("error" to ToolValue.string(toolResult.error ?: "Unknown error")),
            )

        // Build follow-up prompt using C++ (SINGLE SOURCE OF TRUTH)
        val tools = options?.tools ?: ToolRegistry.getAll()
        val toolsPrompt =
            if (options?.keepToolsAvailable == true) {
                CppBridgeToolCalling.formatToolsForPrompt(tools, options.format)
            } else {
                null
            }

        val continuedPrompt =
            CppBridgeToolCalling.buildFollowupPrompt(
                originalPrompt = previousPrompt,
                toolsPrompt = toolsPrompt,
                toolName = toolCall.toolName,
                toolResultJson = resultJson,
                keepToolsAvailable = options?.keepToolsAvailable ?: false,
            )

        val continuationOptions =
            ToolCallingOptions(
                tools = options?.tools,
                maxToolCalls = maxOf(0, (options?.maxToolCalls ?: 5) - 1),
                autoExecute = options?.autoExecute ?: true,
                temperature = options?.temperature,
                maxTokens = options?.maxTokens,
                systemPrompt = options?.systemPrompt,
                replaceSystemPrompt = options?.replaceSystemPrompt ?: false,
                keepToolsAvailable = options?.keepToolsAvailable ?: false,
                format = options?.format ?: ToolCallFormat.Default,
            )

        return generateWithTools(continuedPrompt, continuationOptions)
    }

    // ========================================================================
    // PRIVATE HELPERS
    // ========================================================================

    /**
     * Builds the system prompt with tool definitions using C++ implementation.
     */
    private fun buildToolSystemPrompt(
        tools: List<ToolDefinition>,
        options: ToolCallingOptions,
    ): String {
        // Use C++ implementation for prompt formatting (SINGLE SOURCE OF TRUTH)
        // Pass the format from options to generate model-specific instructions
        val toolsPrompt = CppBridgeToolCalling.formatToolsForPrompt(tools, options.format)

        return when {
            options.replaceSystemPrompt && options.systemPrompt != null -> {
                options.systemPrompt
            }
            options.systemPrompt != null -> {
                "${options.systemPrompt}\n\n$toolsPrompt"
            }
            else -> {
                toolsPrompt
            }
        }
    }

    /**
     * Generate text using streaming and collect all tokens into a single string.
     */
    private suspend fun generateAndCollect(
        prompt: String,
        temperature: Float?,
        maxTokens: Int?,
    ): String {
        val genOptions =
            LLMGenerationOptions(
                maxTokens = maxTokens ?: 1024,
                temperature = temperature ?: 0.7f,
            )

        val tokenFlow = RunAnywhere.generateStream(prompt, genOptions)

        val responseText = StringBuilder()
        tokenFlow.collect { token ->
            responseText.append(token)
        }

        return responseText.toString()
    }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/public/runanywhere_tool_calling.dart
/// Tool Calling Extension for RunAnywhere SDK
///
/// Provides tool calling (function calling) functionality for LLMs.
/// Allows LLMs to request external actions (API calls, device functions, etc.)
///
/// Matches Swift SDK's RunAnywhere+ToolCalling.swift
library runanywhere_tool_calling;

import 'package:runanywhere/foundation/logging/sdk_logger.dart';
import 'package:runanywhere/native/dart_bridge_tool_calling.dart';
import 'package:runanywhere/public/runanywhere.dart';
import 'package:runanywhere/public/types/generation_types.dart';
import 'package:runanywhere/public/types/tool_calling_types.dart';

/// Tool calling extension for RunAnywhere
///
/// Provides:
/// - Tool registration and management
/// - Tool-enabled generation with automatic execution
/// - Manual tool execution support
extension RunAnywhereToolCalling on RunAnywhere {
  // Private static registry - stores tool executors by name
  static final Map<String, ToolExecutor> _toolExecutors = {};
  static final Map<String, ToolDefinition> _toolDefinitions = {};
  static final _logger = SDKLogger('RunAnywhere.ToolCalling');

  // ============================================================================
  // MARK: - Tool Registration
  // ============================================================================

  /// Register a tool with the SDK.
  ///
  /// [definition] Tool definition including name, description, and parameters
  /// [executor] Async function to execute when the tool is called
  ///
  /// Example:
  /// ```dart
  /// RunAnywhereTools.registerTool(
  ///   ToolDefinition(
  ///     name: 'get_weather',
  ///     description: 'Get current weather for a location',
  ///     parameters: [
  ///       ToolParameter(
  ///         name: 'location',
  ///         type: ToolParameterType.string,
  ///         description: 'City name or coordinates',
  ///       ),
  ///     ],
  ///   ),
  ///   (args) async {
  ///     final location = args['location']?.stringValue ?? 'Unknown';
  ///     // Call weather API...
  ///     return {'temperature': NumberToolValue(72), 'condition': StringToolValue('Sunny')};
  ///   },
  /// );
  /// ```
  static void registerTool(ToolDefinition definition, ToolExecutor executor) {
    _toolDefinitions[definition.name] = definition;
    _toolExecutors[definition.name] = executor;
    _logger.info('Registered tool: ${definition.name}');
  }

  /// Unregister a tool by name
  static void unregisterTool(String toolName) {
    _toolDefinitions.remove(toolName);
    _toolExecutors.remove(toolName);
    _logger.info('Unregistered tool: $toolName');
  }

  /// Get all registered tool definitions
  static List<ToolDefinition> getRegisteredTools() {
    return List.unmodifiable(_toolDefinitions.values.toList());
  }

  /// Clear all registered tools
  static void clearTools() {
    _toolDefinitions.clear();
    _toolExecutors.clear();
    _logger.info('Cleared all registered tools');
  }

  // ============================================================================
  // MARK: - Tool Execution
  // ============================================================================

  /// Execute a tool call manually.
  ///
  /// [toolCall] The tool call to execute
  /// Returns ToolResult with success/failure and result data
  static Future<ToolResult> executeTool(ToolCall toolCall) async {
    final executor = _toolExecutors[toolCall.toolName];

    if (executor == null) {
      return ToolResult(
        toolName: toolCall.toolName,
        success: false,
        error: 'Tool not found: ${toolCall.toolName}',
        callId: toolCall.callId,
      );
    }

    try {
      _logger.debug('Executing tool: ${toolCall.toolName}');
      final result = await executor(toolCall.arguments);
      _logger.debug('Tool ${toolCall.toolName} completed successfully');

      return ToolResult(
        toolName: toolCall.toolName,
        success: true,
        result: result,
        callId: toolCall.callId,
      );
    } catch (e) {
      _logger.error('Tool ${toolCall.toolName} failed: $e');
      return ToolResult(
        toolName: toolCall.toolName,
        success: false,
        error: e.toString(),
        callId: toolCall.callId,
      );
    }
  }

  // ============================================================================
  // MARK: - Tool-Enabled Generation
  // ============================================================================

  /// Generate text with tool calling support.
  ///
  /// This is the main entry point for tool-enabled generation.
  /// Handles the full tool calling loop:
  /// 1. Format tools into system prompt
  /// 2. Generate LLM response
  /// 3. Parse tool calls from output
  /// 4. Execute tools (if autoExecute is true)
  /// 5. Continue generation with tool results
  /// 6. Repeat until no more tool calls or max iterations reached
  ///
  /// [prompt] User's question or request
  /// [options] Tool calling options (optional)
  ///
  /// Example:
  /// ```dart
  /// final result = await RunAnywhereTools.generateWithTools(
  ///   'What is the weather in San Francisco?',
  /// );
  /// print(result.text); // "The weather in San Francisco is 72°F and Sunny."
  /// print(result.toolCalls); // [ToolCall(name: 'get_weather', ...)]
  /// ```
  static Future<ToolCallingResult> generateWithTools(
    String prompt, {
    ToolCallingOptions? options,
  }) async {
    final opts = options ?? const ToolCallingOptions();
    final tools = opts.tools ?? getRegisteredTools();
    final formatName = opts.formatName;

    if (tools.isEmpty) {
      // No tools - just do regular generation
      final result = await RunAnywhere.generate(prompt);
      return ToolCallingResult(
        text: result.text,
        toolCalls: [],
        toolResults: [],
        isComplete: true,
      );
    }

    // Build tools JSON
    final toolsJson = toolsToJson(tools);
    _logger.debug('Tools JSON: $toolsJson');
    _logger.debug('Using tool call format: $formatName');

    // Build initial prompt with tools using the specified format
    final toolsPrompt = DartBridgeToolCalling.shared.formatToolsPromptWithFormat(
      toolsJson,
      formatName,
    );
    
    // Build the full prompt with system instructions and user query
    final formattedPrompt = '$toolsPrompt\n\nUser: $prompt';
    _logger.debug('Formatted prompt: ${formattedPrompt.substring(0, formattedPrompt.length.clamp(0, 200))}...');

    // Track all tool calls and results
    final allToolCalls = <ToolCall>[];
    final allToolResults = <ToolResult>[];

    var currentPrompt = formattedPrompt;
    var iterations = 0;
    final maxIterations = opts.maxToolCalls;

    while (iterations < maxIterations) {
      iterations++;

      // Lower temperature for more consistent tool calling behavior
      final genOptions = LLMGenerationOptions(
        maxTokens: opts.maxTokens ?? 1024,
        temperature: opts.temperature ?? 0.3,
      );
      
      // Use streaming like Swift does, then collect all tokens
      final streamResult = await RunAnywhere.generateStream(currentPrompt, options: genOptions);
      final buffer = StringBuffer();
      await for (final token in streamResult.stream) {
        buffer.write(token);
      }
      final responseText = buffer.toString();
      
      _logger.debug('LLM output (iter $iterations): ${responseText.substring(0, responseText.length.clamp(0, 200))}...');

      // Parse for tool calls using C++ bridge (auto-detection like Swift)
      final parseResult = DartBridgeToolCalling.shared.parseToolCall(responseText);

      if (!parseResult.hasToolCall || parseResult.toolName == null) {
        // No tool call - return final result
        return ToolCallingResult(
          text: parseResult.cleanText,
          toolCalls: allToolCalls,
          toolResults: allToolResults,
          isComplete: true,
        );
      }

      // Create tool call
      final toolCall = ToolCall(
        toolName: parseResult.toolName!,
        arguments: parseResult.arguments != null
            ? dynamicMapToToolValueMap(parseResult.arguments!)
            : {},
        callId: parseResult.callId.toString(),
      );
      allToolCalls.add(toolCall);

      _logger.info('Tool call detected: ${toolCall.toolName}');

      if (!opts.autoExecute) {
        // Return for manual execution
        return ToolCallingResult(
          text: parseResult.cleanText,
          toolCalls: allToolCalls,
          toolResults: allToolResults,
          isComplete: false,
        );
      }

      // Execute the tool
      final toolResult = await executeTool(toolCall);
      allToolResults.add(toolResult);

      // Build follow-up prompt with tool result
      final resultJson = toolResult.result != null
          ? toolResultToJsonString(toolResult.result!)
          : '{"error": "${toolResult.error ?? 'Unknown error'}"}';

      currentPrompt = DartBridgeToolCalling.shared.buildFollowupPrompt(
        originalPrompt: prompt,
        toolsPrompt: opts.keepToolsAvailable
            ? DartBridgeToolCalling.shared.formatToolsPrompt(toolsJson)
            : null,
        toolName: toolCall.toolName,
        toolResultJson: resultJson,
        keepToolsAvailable: opts.keepToolsAvailable,
      );

      _logger.debug('Follow-up prompt: ${currentPrompt.substring(0, currentPrompt.length.clamp(0, 200))}...');
    }

    // Max iterations reached - return what we have
    _logger.warning('Max tool call iterations ($maxIterations) reached');
    return ToolCallingResult(
      text: '',
      toolCalls: allToolCalls,
      toolResults: allToolResults,
      isComplete: true,
    );
  }

  /// Continue generation after manual tool execution.
  ///
  /// Use this when autoExecute is false and you've executed tools manually.
  ///
  /// [originalPrompt] The original user prompt
  /// [toolResult] Result from manual tool execution
  /// [options] Tool calling options
  static Future<ToolCallingResult> continueWithToolResult(
    String originalPrompt,
    ToolResult toolResult, {
    ToolCallingOptions? options,
  }) async {
    final opts = options ?? const ToolCallingOptions();
    final tools = opts.tools ?? getRegisteredTools();
    final toolsJson = toolsToJson(tools);

    final resultJson = toolResult.result != null
        ? toolResultToJsonString(toolResult.result!)
        : '{"error": "${toolResult.error ?? 'Unknown error'}"}';

    final followupPrompt = DartBridgeToolCalling.shared.buildFollowupPrompt(
      originalPrompt: originalPrompt,
      toolsPrompt: opts.keepToolsAvailable
          ? DartBridgeToolCalling.shared.formatToolsPrompt(toolsJson)
          : null,
      toolName: toolResult.toolName,
      toolResultJson: resultJson,
      keepToolsAvailable: opts.keepToolsAvailable,
    );

    // Continue with the follow-up
    return generateWithTools(followupPrompt, options: opts);
  }

  // ============================================================================
  // MARK: - Helper Functions
  // ============================================================================

  /// Format tools for system prompt.
  ///
  /// Useful for inspecting or customizing tool prompts.
  static String formatToolsForPrompt([List<ToolDefinition>? tools]) {
    final toolList = tools ?? getRegisteredTools();
    if (toolList.isEmpty) return '';

    final toolsJson = toolsToJson(toolList);
    return DartBridgeToolCalling.shared.formatToolsPrompt(toolsJson);
  }

  /// Parse tool call from LLM output.
  ///
  /// Useful for manual parsing without automatic execution.
  static ToolCall? parseToolCall(String llmOutput) {
    final result = DartBridgeToolCalling.shared.parseToolCall(llmOutput);

    if (!result.hasToolCall || result.toolName == null) {
      return null;
    }

    return ToolCall(
      toolName: result.toolName!,
      arguments: result.arguments != null
          ? dynamicMapToToolValueMap(result.arguments!)
          : {},
      callId: result.callId.toString(),
    );
  }
}

/// Convenience class for tool calling without extension syntax
///
/// Use this for simpler imports:
/// ```dart
/// import 'package:runanywhere/public/runanywhere_tool_calling.dart';
///
/// RunAnywhereTools.registerTool(...);
/// final result = await RunAnywhereTools.generateWithTools('...');
/// ```
class RunAnywhereTools {
  RunAnywhereTools._();

  /// Register a tool with the SDK
  static void registerTool(ToolDefinition definition, ToolExecutor executor) =>
      RunAnywhereToolCalling.registerTool(definition, executor);

  /// Unregister a tool by name
  static void unregisterTool(String toolName) =>
      RunAnywhereToolCalling.unregisterTool(toolName);

  /// Get all registered tool definitions
  static List<ToolDefinition> getRegisteredTools() =>
      RunAnywhereToolCalling.getRegisteredTools();

  /// Clear all registered tools
  static void clearTools() => RunAnywhereToolCalling.clearTools();

  /// Execute a tool call manually
  static Future<ToolResult> executeTool(ToolCall toolCall) =>
      RunAnywhereToolCalling.executeTool(toolCall);

  /// Generate text with tool calling support
  static Future<ToolCallingResult> generateWithTools(
    String prompt, {
    ToolCallingOptions? options,
  }) =>
      RunAnywhereToolCalling.generateWithTools(prompt, options: options);

  /// Continue generation after manual tool execution
  static Future<ToolCallingResult> continueWithToolResult(
    String originalPrompt,
    ToolResult toolResult, {
    ToolCallingOptions? options,
  }) =>
      RunAnywhereToolCalling.continueWithToolResult(
        originalPrompt,
        toolResult,
        options: options,
      );

  /// Format tools for system prompt
  static String formatToolsForPrompt([List<ToolDefinition>? tools]) =>
      RunAnywhereToolCalling.formatToolsForPrompt(tools);

  /// Parse tool call from LLM output
  static ToolCall? parseToolCall(String llmOutput) =>
      RunAnywhereToolCalling.parseToolCall(llmOutput);
}
